CMakeLists.txt

```diff
@@ -4,29 +4,131 @@
 
 cmake_minimum_required(VERSION 2.8)
 
-if(DEFINED CMAKE_BUILD_TYPE)
-   set(CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE} CACHE STRING "Debug Release RelWithDebInfo MinSizeRel")
-else()
-   set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Debug Release RelWithDebInfo MinSizeRel")
+project(uavcan C CXX)
+
+#=============================================================================
+# build options
+#
+# options are listed in a table format below
+set(opts
+    # name: type: default value: string options list : description
+    "CMAKE_BUILD_TYPE:STRING:RelWithDebInfo:Debug Release RelWithDebInfo MinSizeRel:Build type."
+    "CMAKE_CXX_FLAGS:STRING:::C++ flags."
+    "CMAKE_C_FLAGS:STRING:::C flags."
+    "UAVCAN_USE_CPP03:BOOL:OFF::Use cpp03 standard."
+    "UAVCAN_NO_ASSERTIONS:BOOL:OFF::Enable assertions."
+    "UAVCAN_USE_EXTERNAL_SNPRINTF:BOOL:OFF::Use external snprintf."
+    "UAVCAN_MEM_POOL_BLOCK_SIZE:STRING:48::memory pool block size"
+    "UAVCAN_STM32_TIMER_NUMBER:STRING:5::default timer"
+    "UAVCAN_MAX_NETWORK_SIZE_HINT:STRING:16::max network size"
+    "UAVCAN_PLATFORM:STRING:generic:generic linux stm32:Platform."
+    "UAVCAN_OS:STRING:generic:generic posix nuttx chibios:Operating system."
+    "UAVCAN_STM32_NUM_IFACES:STRING:2:1 2:Number of interfaces."
+    )
```

Comment:
The question is, can we tolerate extra maintenance requirements for the
sake of marginal improvement of user experience. Keeping in mind that all
of the relevant build config options are already documented here
http://uavcan.org/Implementations/Libuavcan/Tutorials/1._Library_build_configuration/,
I'm inclined to say that no, we can't.
On Oct 4, 2015 8:32 PM, "James Goppert" <notifications@github.com> wrote:

> In CMakeLists.txt
> <https://github.com/UAVCAN/libuavcan/pull/67#discussion_r41100037>:
>
> > +# options are listed in a table format below
> > +set(opts
> > +    # name: type: default value: string options list : description
> > +    "CMAKE_BUILD_TYPE:STRING:RelWithDebInfo:Debug Release RelWithDebInfo MinSizeRel:Build type."
> > +    "CMAKE_CXX_FLAGS:STRING:::C++ flags."
> > +    "CMAKE_C_FLAGS:STRING:::C flags."
> > +    "UAVCAN_USE_CPP03:BOOL:OFF::Use cpp03 standard."
> > +    "UAVCAN_NO_ASSERTIONS:BOOL:OFF::Enable assertions."
> > +    "UAVCAN_USE_EXTERNAL_SNPRINTF:BOOL:OFF::Use external snprintf."
> > +    "UAVCAN_MEM_POOL_BLOCK_SIZE:STRING:48::memory pool block size"
> > +    "UAVCAN_STM32_TIMER_NUMBER:STRING:5::default timer"
> > +    "UAVCAN_MAX_NETWORK_SIZE_HINT:STRING:16::max network size"
> > +    "UAVCAN_PLATFORM:STRING:generic:generic linux stm32:Platform."
> > +    "UAVCAN_OS:STRING:generic:generic posix nuttx chibios:Operating system."
> > +    "UAVCAN_STM32_NUM_IFACES:STRING:2:1 2:Number of interfaces."
> > +    )
>
> I believe exposing the options to cmake/ the cmake-gui is important for
> end users. The typical way is to then use the cmake flags to drive a
> config.h.in -> config.h file using cmakedefine. If you note, when running
> the cmake-gui, this presents the user with a nice dialog with check boxes,
> and drop down lists for options.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/UAVCAN/libuavcan/pull/67/files#r41100037>.
>